sync policy, MIME override, encoding support, and more

Bhpsngum opened the pull request:
## Synchronous policy and fixes
- Add new non-standard option `syncPolicy: "warn" | "enabled" | "disabled"` (default is "warn") to explicitly enable/disable synchronous XHR with or without warning
	- `"warn"`: enable synchronous xhr but shows a warning when the user attempts to call `xhr.open()` with sync flag:
		> [Deprecation] Synchronous XMLHttpRequest is deprecated because of its detrimental effects to the end user's experience. For more information, see https://xhr.spec.whatwg.org/#sync-flag
	- `"enabled"`: enable synchronous xhr, no warnings
	- `"disabled"`: disable synchronous, calling `xhr.open()` with sync flag won't show any warning, but an exception is thrown when calling `xhr.send()` afterwards:
		> Synchronous requests are disabled for this instance.
- Fix problem with data not encoded properly when using synchronous request (#29)
- Sync responses now write data into OS temporary directory instead of cwd (since rwx permissions can be altered for cwd)
- If file system operations fail, an error message will be thrown:
	> Synchronous operation aborted: Unable to access the OS temporary directory for read/write operations.
## Encoding and MIME type support
- Add `xhr.overrideMimeType(type)` (#27)
- Add support for encodings other than utf-8 (can be read from `Content-Type` header or through user's custom encoding via `xhr.overrideMimeType()`)
## Non-standard methods
- Removed `xhr.setDisableHeaderCheck()` and instead added a non-standard `disableHeaderCheck: true | false` (default is `false`)
- Add deprecation message for `xhr.getRequestHeader()`:
	> `xhr.getRequestHeader()` is deprecated and will be removed in a future release. It’s non-standard and not part of the XHR spec.
## Others
- Prevent `XMLHttpRequest` to be called without `new` operator
- Unifies `createFileOrSyncResponse` and `createResponse` into one method `createResponse`
- Brings some methods back to private scope to prevent tampering (especially the `readyState` value)
- Add immutable flags value UNSENT, OPENED, HEADERS_RECEIVED, LOADING, and DONE to both `XMLHttpRequest` and any newly-created xhr object
- Add two more options, `textDecoder: <func>` for users to specify custom text decoder (to support more range of encodings for old Node versions) and `xmlParser: <func>` to support parsing for `xhr.responseXML`

Bhpsngum committed:
[draft] syncPolicy, MIME support, and more

Bhpsngum committed:
make disableHeaderCheck an option and add deprecation for xhr.getRequestHeader

Bhpsngum committed:
change fs error msg

Bhpsngum committed:
change to shorter message

Bhpsngum committed:
check which version does not support iso-8859-1

Bhpsngum committed:
add dispatch

Bhpsngum committed:
fix typo

Bhpsngum committed:
matrix not fail fast

Bhpsngum commented:
A few note on this: Since `TextDecoder` only adds support for more encodings on later node versions after v12, the test on v12 failed

We can either increase the minimum node version limit or change to use an external dependency for decoding (such as `iconv-lite`), there is a new option called `textEncoder` so user can pass their custom decoder into.

Bhpsngum committed:
Add barebone support for extensive text decoding and xml parsing

Bhpsngum committed:
Fix docs typo

Bhpsngum committed:
Fix test of mime type

Bhpsngum committed:
Fix test of mime type (failsafe)

mjwwit commented:
At first glance this looks really solid, thanks for the great work! :heart:
I'll have a more detailed look in the coming days, but we can probably get this merged this week.

> We can either increase the minimum node version limit or change to use an external dependency for decoding (such as `iconv-lite`), there is a new option called `textEncoder` so user can pass their custom decoder into.

I have no problems raising the minimum version to 13 or 14.

Bhpsngum commented:
I have no idea if we can keep the minimum version as v12, with caveats that encoding won't work with many options and need to specify a custom decoder (`textDecoder: function (buf, enc) { return iconv.decode(buf, enc) }` for example), or we raise it to the version where it already supports more options.

Please let me know what you think about this.

mjwwit commented:
> I have no idea if we can keep the minimum version as v12, with caveats that encoding won't work with many options and need to specify a custom decoder (`textDecoder: function (buf, enc) { return iconv.decode(buf, enc) }` for example), or we raise it to the version where it already supports more options.
> 
> Please let me know what you think about this.

Keeping the minimum version with caveats is probably a bad idea. People will probably not read about these caveats proactively, and the package won't be fully functional on Node.js 12. I'll make the next release a major. People can still choose not to upgrade because of that.

mjwwit commented on README.md:
```diff
@@ -13,19 +13,94 @@ XHR object.
+			<td><code>false</code></td>
+			<td>Set to any truthy value to prevent a process from not exiting naturally because a request socket from this module is still open</td>
+        </tr>
+		<tr>
+            <td><code>maxRedirects</code></td>
+			<td><code>20</code></td>
+			<td>Allows control over the maximum number of redirects that are followed</td>
+        </tr>
+		<tr>
+            <td><code>allowFileSystemResources</code></td>
+			<td><code>true</code></td>
+			<td>Allows control access to the local filesystem through the <code>file:</code> protocol</td>
+        </tr>
+		<tr>
+            <td><code>origin</code></td>
+			<td><code>undefined</code></td>
+			<td>Set a base URL for the requests called using this instance. The resulting request URL will be constructed as follows: <code>new URL(url, origin)</code></td>
+        </tr>
+		<tr>
+            <td><code>syncPolicy</code></td>
+			<td><code>"warn"</code></td>
+			<td>Control feature behavior of the synchronous feature:<ul><li><code>"disabled"</code>: Disable the feature completely, throws error after calling <code>.send()</code> if in synchronous mode</li><li><code>"warn"</code>: Enable the feature, but show a warning when calling <code>.open()</code> with synchronous mode</li><li><code>"enabled"</code>: Enable the feature without showing any additiona ...
+        </tr>
+		<tr>
+            <td><code>disableHeaderCheck</code></td>
+			<td><code>false</code></td>
+			<td>Disable the check against forbidden headers to be added to a XHR request</td>
+        </tr>
+		<tr>
+            <td><code>xmlParser</code></td>
+			<td>none</td>
+			<td>Specify a parser (non-async) to parse document from text when <code>xhr.responseType</code> is either <code>"document"</code> or in text format. If the parser is invalid or omitted, <code>xhr.responseXML</code> will be <code>null</code></td>
+        </tr>
+		<tr>
+            <td><code>textDecoder</code></td>
+			<td><code>TextDecoder</code> or <code>buf.toString(enc)</code> depending on Node version</td>
+			<td>Specify a text decoder (non-async), accepting a buffer <code>buf</code> and encoding <code>enc</code> to decode to desired encoding.<br>Note that <code>TextDecoder</code> at version 12 does not support a wide range of encodings than later node version does</td>
+        </tr>
+    </tbody>
+</table>
```
Github actually supports tables in markdown: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-tables. This would probably greatly enhance readability of the source file.

Bhpsngum replied:
problem is the colspan and rowspan in the table i specified to group similar descriptions, which GitHub Markdown doesn't support yet :(

mjwwit replied:
Ah that sucks.. In that case we'll keep it like this.

Bhpsngum commented:
alright, that sounds good for me too 👍

mjwwit commented on lib/XMLHttpRequest.js:
```diff
@@ -439,14 +562,18 @@ function XMLHttpRequest(opts) {
 
+    if (opts.syncPolicy === "disabled") {
+      throw new Error("Synchronous requests are disabled for this instance.");
+    }
```
But this could still be an async XHR at this point right? Maybe the condition should be amended with `&& !settings.async`.

Bhpsngum replied:
yeah i forgot, thanks for the fixes

mjwwit commented on tests/test-headers.js:
```diff
@@ -10,7 +10,7 @@ var server = http.createServer(function (req, res) {
   assert.equal("node-XMLHttpRequest-test", req.headers["user-agent"]);
   // Test header set with blacklist disabled
-  assert.equal("http://github.com", req.headers["referer"]);
+  // assert.equal("http://github.com", req.headers["referer"]);
```
We should either remove these assertions or re-enable them

Bhpsngum replied:
i need to rewrite this test with our new `disableHeaderCheck` flag added, i forgot about that 😅

mjwwit reviewed (changes requested):
Looking good! The `syncPolicy` `"disabled"` behavior contains a bug though; when sending an async request it still throws an error. Maybe this would also be a good candidate for a test.

Bhpsngum committed:
bump minimum node version to v13, fix syncPolicy for non-async and add basic anti-pollution measure

Bhpsngum commented:
I've fixed the problems you addressed, and went further and added a basic measure to anti-prototype pollution, there is also a new test i added for that purpose (`test-pollution.js`).

mjwwit commented:
It looks complete now, I'll run some usage tests to see if I can discover any problems. I'm not expecting to find anything, so I'll probably merge quickly after that. The release will follow shortly after that, probably on the same day.

Bhpsngum commented:
Alright, all sounds good to me 👍

mjwwit commented:
Thanks again for the solid work :heart:. I've released it as version 4.0.0.
